url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/environment/geometry/ParametricCurve.java

```diff
@@ -0,0 +1,633 @@
+        result = result.add(p.multiply(vt));
+
+        return result;
+    }
+
+    /**
+    Note that current implemented quadratic spline interpolation strategy
+    is to first express the second order quadric as a third order Bezier
+    spline, and later evaluate the Bezier spline, as noted on [.FONT2006].
+
+    @todo Express this interpolation as a QUAD_MATRIX and similar scheme to
+    other points in this class.
+    */
+    private Vector3D evaluateQuadratic(int nseg, double t)
+    {
+        Vector3D[] startingSegmentControl = points.get(nseg - 1);
+        Vector3D[] endingSegmentControl = points.get(nseg);
+        Vector3D p;
+
+        Vector3D qp0, qp1, qp2;
+        qp0 = startingSegmentControl[0];
+        qp1 = endingSegmentControl[1];
+        qp2 = endingSegmentControl[0];
+
+        // p0
+        Vector3D result = qp0;
+        double vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += BEZIER_MATRIX.M[i][0] * (Math.pow(t, 3 - i));
+        }
+        result = result.multiply(vt);
+
+        // p1
+        vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += BEZIER_MATRIX.M[i][1] * (Math.pow(t, 3 - i));
+        }
+        p = qp0.add((qp1.substract(qp0)).multiply(2/3));
+        result = result.add(p.multiply(vt));
+
+        // p2
+        vt = 0;
+
+        for (int i = 0; i < 4; i++) {
+            vt += BEZIER_MATRIX.M[i][2] * (Math.pow(t, 3 - i));
+        }
+        p = qp1.add((qp2.substract(qp0)).multiply(1/3));
+        result = result.add(p.multiply(vt));
+
+        // p3
+        vt = 0;
+        for (int i = 0; i < 4; i++) {
+            vt += BEZIER_MATRIX.M[i][3] * (Math.pow(t, 3 - i));
+        }
+        p = endingSegmentControl[0];
+        result = result.add(p.multiply(vt));
+
+        return result;
+    }
+
+    private Vector3D evaluateHermite(int nseg, double t) {
+        Vector3D[] startingSegmentControl = points.get(nseg - 1);
+        Vector3D[] endingSegmentControl = points.get(nseg);
+
+        // p1
+        Vector3D result = startingSegmentControl[0];
+        double vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += HERMITE_MATRIX.M[i][0] * (Math.pow(t, 3 - i));
+        }
+        result = result.multiply(vt);
+
+        // p4
+        vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += HERMITE_MATRIX.M[i][1] * (Math.pow(t, 3 - i));
+        }
+        Vector3D p = endingSegmentControl[0];
+        result = result.add(p.multiply(vt));
+
+        // r1
+        vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += HERMITE_MATRIX.M[i][2] * (Math.pow(t, 3 - i));
+        }
+        p = startingSegmentControl[2];
+        result = result.add(p.multiply(vt));
+
+        // r4
+        vt = 0;
+        for ( int i = 0; i < 4; i++ ) {
+            vt += HERMITE_MATRIX.M[i][3] * (Math.pow(t, 3 - i));
+        }
+        p = endingSegmentControl[1];
+        result = result.add(p.multiply(vt));
+
+        return result;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 398:433, 437:474.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.